wasm/section.go

```diff
@@ -0,0 +1,696 @@
+type SectionCode struct {
+	Section
+	Bodies []FunctionBody
+}
+
+func (m *Module) readSectionCode(r io.Reader) error {
+	s := &SectionCode{}
+
+	count, err := leb128.ReadVarUint32(r)
+	if err != nil {
+		return err
+	}
+	s.Bodies = make([]FunctionBody, count)
+	logger.Printf("%d function bodies\n", count)
+
+	for i := uint32(0); i < count; i++ {
+		logger.Printf("Reading function %d\n", i)
+		if s.Bodies[i], err = readFunctionBody(r); err != nil {
+			return err
+		}
+		s.Bodies[i].Module = m
+
+	}
+
+	m.Code = s
+	if m.Function == nil || len(m.Function.Types) == 0 {
+		return errors.New("Missing section: function")
+	}
+	if len(m.Function.Types) != len(s.Bodies) {
+		return errors.New("The number of entries in the function and code section are unequal")
+	}
+
+	if m.Types == nil {
+		return errors.New("Missing section: type")
+	}
+
+	for _, index := range m.Function.Types {
+		if int(index) >= len(m.Types.Entries) {
+			return errors.New("Invalid index")
+		}
+
+		m.Types.Entries[index].body = &s.Bodies[index]
+	}
+
+	return nil
+}
+
+type FunctionBody struct {
+	Module *Module // The parent module containing this function body, for execution purposes
+	Locals []LocalEntry
+	Code   []byte
+}
+
+func readFunctionBody(r io.Reader) (FunctionBody, error) {
+	f := FunctionBody{}
+
+	bodySize, err := leb128.ReadVarUint32(r)
+	if err != nil {
+		return f, err
+	}
+
+	r = io.LimitReader(r, int64(bodySize))
+
+	localCount, err := leb128.ReadVarUint32(r)
+	if err != nil {
+		return f, err
+	}
+	f.Locals = make([]LocalEntry, localCount)
+
+	for i := uint32(0); i < localCount; i++ {
+		if f.Locals[i], err = readLocalEntry(r); err != nil {
+			return f, err
+		}
+	}
+
+	logger.Printf("bodySize: %d, localCount: %d\n", bodySize, localCount)
+
+	code, err := ioutil.ReadAll(r)
+
+	if code[len(code)-1] != end {
+		return f, errors.New("Function body does not end with 0x0b (end)")
+	}
+
+	f.Code = code[:len(code)-1]
+	if err != nil {
+		return f, err
+	}
+
+	return f, nil
+}
+
+type LocalEntry struct {
+	Count uint32
+	Type  ValueType
+}
+
+func readLocalEntry(r io.Reader) (LocalEntry, error) {
+	l := LocalEntry{}
+	var err error
+
+	l.Count, err = leb128.ReadVarUint32(r)
+	if err != nil {
+		return l, err
+	}
+
+	l.Type, err = readValueType(r)
+	if err != nil {
+		return l, err
+	}
+
+	return l, nil
+}
+
+// SectionData describes the intial values of a module's linear memory
+type SectionData struct {
+	Section
+	Entries []DataSegment
+}
+
+func (m *Module) readSectionData(r io.Reader) error {
+	s := &SectionData{}
+	count, err := leb128.ReadVarUint32(r)
+	if err != nil {
+		return err
+	}
+
+	s.Entries = make([]DataSegment, count)
+
+	for i := uint32(0); i < count; i++ {
+		if s.Entries[i], err = readDataSegment(r); err != nil {
+			return err
+		}
+	}
+
+	m.Data = s
+	return err
+}
+
+type DataSegment struct {
+	Index uint32
+	// should return i32, computes the offset at which to place the elements
```

Comment:
Yes, that's due to the encoding for [`elem_segment`](https://github.com/WebAssembly/design/blob/master/BinaryEncoding.md#element-section) and [`data_segment`]([url](https://github.com/WebAssembly/design/blob/master/BinaryEncoding.md#data-section)) being the same.